src/csharp/Microsoft.Spark/Sql/CastUnpickledItems.cs

```diff
@@ -0,0 +1,94 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System;
+using System.Collections;
+using System.Collections.Generic;
+using System.Linq;
+using Microsoft.Spark.Sql.Types;
+
+namespace Microsoft.Spark.Sql
+{
+    /// <summary>
+    /// Cast UnpickledItems.
+    /// </summary>
+    internal static class CastUnpickledItems
+    {
+        /// <summary>
+        /// Cast unpickledItems from arraylist to the appropriate array.
+        /// </summary>
+        /// <param name="unpickledItems">Unpickled objects to be cast as necessary.</param>
+        /// <returns>Unpickled objects after casting.</returns>
+        public static object[] Cast(object unpickledItems)
+        {
+            var castUnpickledItems = new List<object>();
+            foreach (object obj in (object[])unpickledItems)
+            {
+                castUnpickledItems.Add(
+                    (obj.GetType() == typeof(RowConstructor)) ? obj : CastArray(obj));
+            }
+
+            return castUnpickledItems.ToArray();
+        }
+
+        /// <summary>
+        /// Cast simple array and array of arrays.
+        /// </summary>
+        /// <param name="obj">object to be cast as necessary.</param>
+        /// <returns>Typed array after casting.</returns>
+        public static object CastArray(object obj)
+        {
+            if (obj is object[] objArr)
```

Comment:
does this work with `Row[]`?

Reply:
From what I have observed, if udf takes `Row[]` like the example [here](https://github.com/dotnet/spark/blob/25abbfa90357b658c23db29e08cacb67b82361a1/src/csharp/Microsoft.Spark.E2ETest/UdfTests/UdfComplexTypesTests.cs#L73). It will be `RowConstructor`, and [`FromInternal`](https://github.com/dotnet/spark/blob/25abbfa90357b658c23db29e08cacb67b82361a1/src/csharp/Microsoft.Spark/Sql/Types/ComplexTypes.cs#L299) should handle such cases.